### ISSUE TITLE
Read router.payload from Switch and its child Components

### ISSUE THREAD

**Issue by bultas** [faceyspacey/redux-first-router#149]
Given I have routes:

```
routeA /
routeB /:variable
```

and *SwitchComponent* which is connected to router.type

```
const SwitchComponent = ({ type }) => {
  if (type === routeA) return <ComponentA>
  if (type === routeB) return <ComponentB>
}
```

and *ComponentB* is connected to router.payload { variable }
+*variable* is its required prop
(btw i personally prefer use another name then *location* for router reducer because *location* is global variable and may result to some misunderstandings/bugs in code)

```
const ComponentB = ({ variable }) => (
  doSomethingWith(variable)
)
```

**Problem:**

* Given that init load is RouteB.
* Then I click to redirect to RouteA
* Then connected *ComponentB* inside RouteB have problem because is still mounted and connected to payload which is empty now (because of dispatched location change).

Interesting thing is that if:

* Init load is RouteA
* Then I click to redirect to RouteB
* Then I click to redirect to RouteA
* Then everything is ok, because connect(selector) in *SwitchComponent* is called before connect(selector) in *ComponentB* so *ComponentB* is already unmounted. = no problem with undefined variable in render.

Similar behaviour/problem is with browser native back button.

* Init load is RouteB
* in browser history is previous route RouteA
* use native browser back button
* redirect to RouteA but with slightly different conditions then if I use dispatch location change (order of connect(selectors) is different)
   

**Questions:**

Is this behaviour expected?
Is this correct way to handle Component Switch?
Is good practice to read payload information inside *SwitchComponent* child or I have to pass them from one main component or handle exceptions in each component which will be reading directly from Router State? (it's not something what I want)

Thank you for advice.

**Comment by faceyspacey** [comment]
I'm not fully understanding the problem, but it sounds to me like you have a child component connected to `payload.someVariable` and this component isn't supposed to show after the route changes. I'm not sure why this is happening--basically the component shouldn't show if the `type` or your other checks change. It's fine to check payload key/vals to determine what component to show. Occasionally that can be the precise perfect way to handle your application-specific problem. 

Bottom line is I can't understand why you are having this problem if you have this code:

```js
const SwitchComponent = ({ type }) => {
  if (type === routeA) return <ComponentA>
  if (type === routeB) return <ComponentB>
}
```

Only one component can show, and `react-redux` is rendered from the top down--so children components won't re-render until this resolves. So maybe I'm missing something.

**Comment by bultas** [comment]
Thanks for reply.. 
I wanted simulate the problem, so I did repo https://github.com/bultas/redux-first-router-problem, where I found the bug in version `"react-redux": "^4.4.5"` which was somehow fixed in version `"^5.0.6"`.

If you are interested about problem,.. You can reproduce it with:

1. clone the repo (https://github.com/bultas/redux-first-router-problem)
2. npm install
3. npm start
4. start own server to server index.html (I just used http-server)
5. go to http://localhost:8080/#/some
6. click on the button
7. You will see prop validation problem (connect order problem)
